replica/server.go

```diff
@@ -354,25 +380,60 @@ func (s *Server) Restore(url, name string) (err error) {
 		return err
 	}
 
-	if err := backupstore.RestoreDeltaBlockBackup(backupURL, toFile); err != nil {
-		return err
-	}
+	restoreObj := NewRestore(toFile)
 
-	newDisk := disk{
-		Parent:      "",
-		Name:        snapName,
-		Removed:     false,
-		UserCreated: false,
-		Created:     util.Now(),
+	config := &backupstore.DeltaRestoreConfig{
+		BackupURL: backupURL,
+		DeltaOps:  restoreObj,
+		Filename:  toFile,
 	}
 
-	if err := r.encodeToFile(newDisk, snapName+metadataSuffix); err != nil {
+	if err := backupstore.RestoreDeltaBlockBackup(config); err != nil {
+		logrus.Errorf("failed to initiate Restore backup for url[%v]: %v", backupURL, err)
 		return err
 	}
 
-	r, err = r.Revert(snapName, util.Now())
-	if err != nil {
-		return err
-	}
-	return r.Close()
+	var (
+		restoreProgress int
+		restoreError    error
+	)
+	periodicChecker := time.NewTicker(PeriodicRefreshIntervalInSeconds * time.Second)
+	go func() {
+		for t := range periodicChecker.C {
+			logrus.Debugf("checking restore progress at %v", t)
+			restoreObj.Lock()
+			restoreProgress = restoreObj.RestoreProgress
+			restoreError = restoreObj.RestoreError
+			lastUpdate := restoreObj.LastUpdatedAt
+			restoreObj.Unlock()
+
+			if restoreProgress == 100 {
+				break
+			}
+			if restoreError != nil {
+				break
+			}
+			now := time.Now()
+			diff := now.Sub(lastUpdate)
+			restoreObj.Unlock()
+
+			if diff.Minutes() > ProgressBasedTimeoutInMinutes {
+				logrus.Errorf("no restore update happened since %v minutes. Returning failure",
+					ProgressBasedTimeoutInMinutes)
+				periodicChecker.Stop()
+				return
+			}
+		}
+		periodicChecker.Stop()
+
+		if restoreProgress == 100 {
```

Comment:
https://github.com/longhorn/backupstore/blob/master//deltablock.go#L378
I'm setting the progress to 100 only after successful completion of restore. The round error you might be referring to wouldn't occur in this scneario because I have split the progress into 95%(for restore completion) and 5% for post restore operation. If the progress is 100, it implies restoration is complete